crates/ruff_server/src/session.rs

```diff
@@ -232,46 +236,55 @@ impl Workspaces {
 
     fn snapshot(&self, document_url: &Url) -> Option<DocumentRef> {
         self.workspace_for_url(document_url)
-            .and_then(|w| w.open_documents.snapshot(document_url))
+            .and_then(|(_, workspace)| workspace.open_documents.snapshot(document_url))
     }
 
     fn controller(&mut self, document_url: &Url) -> Option<&mut DocumentController> {
         self.workspace_for_url_mut(document_url)
-            .and_then(|w| w.open_documents.controller(document_url))
+            .and_then(|(_, workspace)| workspace.open_documents.controller(document_url))
     }
 
     fn configuration(&self, document_url: &Url) -> Option<&Arc<RuffConfiguration>> {
         self.workspace_for_url(document_url)
-            .map(|w| &w.configuration)
+            .map(|(_, workspace)| &workspace.configuration)
+    }
+
+    fn reload_configuration(&mut self, changed_url: &Url) -> crate::Result<()> {
+        let (path, workspace) = self
+            .workspace_for_url_mut(changed_url)
+            .ok_or_else(|| anyhow!("Workspace not found for {changed_url}"))?;
+        workspace.reload_configuration(path);
+        Ok(())
     }
 
     fn open(&mut self, url: &Url, contents: String, version: DocumentVersion) {
-        if let Some(w) = self.workspace_for_url_mut(url) {
-            w.open_documents.open(url, contents, version);
+        if let Some((_, workspace)) = self.workspace_for_url_mut(url) {
+            workspace.open_documents.open(url, contents, version);
         }
     }
 
     fn close(&mut self, url: &Url) -> crate::Result<()> {
         self.workspace_for_url_mut(url)
             .ok_or_else(|| anyhow!("Workspace not found for {url}"))?
+            .1
             .open_documents
             .close(url)
     }
 
-    fn workspace_for_url(&self, url: &Url) -> Option<&Workspace> {
+    fn workspace_for_url(&self, url: &Url) -> Option<(&Path, &Workspace)> {
```

Comment:
Nit: I couldn't guess the return value of `workspace_for_url` from the call site (I had no idea what the first tuple element should be). I had to navigate to the method declaration and read the code. 

Maybe rename to `entry_for_url` and `entry_for_url_mut` but keep `workspace_for_url` and `workspace_for_url_mut` that use `entry_for_url` internally. I think having `workspace_for_url` and `workspace_for_url_mut` that returns a `&Workspace` (or `&mut Workspace`) does make sense and avoids destructuring or indexing in some call sites.

Reply:
Okay! I was worried that having separate methods would be 'too much' but I think we can justify it here.